### DISCUSSION TITLE
Welcome to Harmony-of-the-Spheres Discussions!

### DISCUSSION THREAD

**Discussion by TheHappyKoala** [TheHappyKoala/Harmony-of-the-Spheres#130] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by TheHappyKoala** [comment]
@HugoGranstrom 

Finally! We won't have to shit post in odd closed pull request threads or issues 🤣

**Comment by HugoGranstrom** [comment]
Hahahaha wonderful! 🤣🚀 Oh so I have longed for this moment 😛 We can simulate the true experience though by naming threads names that could have been a PR och issue B)

Hahahahah 🤣 🤣 🤣 🤣 That's what I call poetry! (Now I know two parts of it, this quote and the 42. I haven't read it tbh 😅)

Great! Can we expect to see Material Design?😝🤣 Yeah feels like I'm waiting half the time for the wiki to load so it's a good call 👍 I must say that at first the wrapping of the data in text was quite nice but when I'm on my 10th scenario it has gotten a bit too repeating, so I'm 50/50 in that case.

100 visitors per day isn't bad at all :o That's quite a lot! 🚀 And burst of 3000 just shows what a bit of publicity can do :D. 
Wow, someone using HOS in their teaching 🤩 that's the coolest of them all. But it also puts on a bit of pressure to not mess things up if some school kid somewhere can't do their homework on how Omouamoua moved through the solar system 😮

Julmust sounds safer than coding in a hat that can blow off at anytime 🤣 

Yes, it definitely helps in expanding one's coding toolbox. I'm doing it in Nim so I've got to learn a lot about its standard library and string parsing so far. And today we had to build a very simple VM to solve the task which I haven't really done before but it was a lot of fun 😄 Talking about doing things you don't normally do, is the next step for me to learn how to do CSS properly? 😨

I'm not watching much in particular, just what's on at the moment and hockey of course. Otherwise, it's mostly youtube edutainment videos that on my list. Prime haven't mad much notice here I Sweden compare to Netflix and the others but perhaps that changing now that they've established themselves here? Let's just hope they have better subtitles than translations of their items 🤣

**Comment by TheHappyKoala** [comment]
Happy 2021!!! 

Had a good start to what will hopefully be a better year than the previous one??? 

I've been working on the spaceflight stuff and have made some solid progress, but I've run into some weird bug where, it only happens sometimes, when the spacecraft enters another sphere of influence, everything goes NaN. Been trying to figure out why it only happens in some cases, but so far no luck, so I'm gonna leave it for a week or so and see if something comes to mind, because I'm starting to go proper 🤪 . 

Other than that things are kicking... Nice to be working again and spring's just around the corner, so yay! 

As for learning CSS, I wouldn't wish that on my worst enemy 😝 . It's useful, sure, and it's not as bad as it used to be, but I just find writing clean and sensible CSS to be nigh on impossible. Then again the learning curve is really shallow, so if you want to animate a box or two you can learn how to do so in like ten minutes. 

Gonna make myself some new exoplanet scenarios now... All I have to do is "npm run exoplanets" and voila, done. That should restore my programming confidence 🤣 . 

Hope school is fun and that you're happy and kicking!

**Comment by HugoGranstrom** [comment]
Happy 2021 to you too!! 🎊 It's been fine. I'm reading some really interesting courses right now, one in machine learning, one on Computational science (solving PDEs specifically), and one on thermodynamics. They are a blast and has lots of programming in them 😄 Nice to hear that things are kicking for you! I'm quite fond of the winter tbh. It's so bright and not gray and wet 😉 

That spaceflight bug seems really odd 🤔 From what I can recall, we don't use the orbital elements from the last timestep in the current one so we shouldn't get some strange mix of the two. Hmm... Yeah leaving something when it doesn't work is probably a good idea. It can really drive you mad trying to find what causes it. And in the end, it was all just because you used the wrong variable in some part of the code you didn't look at  at all 🤣 

🤣🤣 true true! Yeah if you are writing it yourself from scratch it may actually be maintainable, but anyone else looking at it will go 😱 and run away from their computer at at least twice the speed of light! 🤣 

Hahaha 🤣 sounds really hard!

**Comment by TheHappyKoala** [comment]
Done a bit of digging regarding the spaceflight bug and it seems like what happens sometimes, but not every time, when the spaceship enters another sphere of influence is that the eccentric and true anomalies come out as NaN, which throws a spanner in the calculations... Any what might be happening here? 

Oh, and happy last day of winter!

**Comment by HugoGranstrom** [comment]
That's a good lead! The eccentric (and mean) anomalies are calculated from the true anomaly so I would guess it the calculation of the true anomaly that goes wrong sometimes. When you say that it sometimes does it, do you mean in some scenarios but not others, or do you mean that for the same scenario given the exact same parameters (dt etc) you only sometimes get NaN? Would be great with a reproducible scenario 😄 

My theory: [here](https://github.com/TheHappyKoala/Harmony-of-the-Spheres/blob/92aac452c832f8dab4228be0d7f664be0e7a9df6/src/physics/spacecraft/lambert.ts#L399) we calculate the true anomaly using `acos` which returns nan if the argument isn't within [-1, 1]. So if we are really close to 1 for example it could be float stuff that makes it 1.000000000000000000000000000000000000000000000000000000000000000001 or something that returns NaN. IF this is the case we could solve it by applying min(max()) to the argument beforehand.

**Comment by HugoGranstrom** [comment]
I have no idea why it only happens when entering a new sphere of influence though :/ Even if the old mass of influence was used somehow it should affect it that much, it would just continue on its old path, right?

Same to you 🌸 The sun and the heat is nice 😄

**Comment by TheHappyKoala** [comment]
I can reproduce it, but it only happens for planets with a smaller soi, so it never happens when I go to Jupiter, Saturn or even Neptune... Getting rather close now, though, as I've pinned it down to the solveKeplerEq method... Fingers crossed 😅

**Comment by TheHappyKoala** [comment]
In the while loop, when the eccentricity is greater than 1, E has a value of -6305.985121061316 and when you try to Math.cosh(E) and Math.sinh(E) you get NaN

**Comment by TheHappyKoala** [comment]
So your hypothesis was spot on

**Comment by TheHappyKoala** [comment]
Now let's see if clamping these guys will fix the issue...

**Comment by TheHappyKoala** [comment]
Noops... That did not work 🤣

**Comment by HugoGranstrom** [comment]
Ahh I see so sinh(E) and cosh(E) becomes Infinity and when we divide infinity by infinity we get NaN 🤯

**Comment by HugoGranstrom** [comment]
So our problem is that the solver doesn't converge for hyperbolic orbits very close to a body? And we every only get so close to the really small ones. Have I understood this correctly?

**Comment by TheHappyKoala** [comment]
Yups... Shenanigans

**Comment by TheHappyKoala** [comment]
We're gonna solve this

**Reply by TheHappyKoala** [reply]
What is the significance of E when a trajectory is hyperbolic???

**Reply by TheHappyKoala** [reply]
@HugoGranstrom Ooops, replied to myself

**Reply by HugoGranstrom** [reply]
I get notified either way 😉 

I don't know what significance it has really but we do use it to calculate the true anomaly for both elliptic and hyperbolic orbits so we can't get around it I think. Unless we find other equations for propagating hyperbolic orbital elements

**Reply by TheHappyKoala** [reply]
I'm like a caricature of an old person using "new technology"

**Reply by HugoGranstrom** [reply]
Haha 🤣 You're the one who found this discussion feature first though. Had no idea it existed before you showed me

**Reply by HugoGranstrom** [reply]
I guess we could use a bracketing method instead of Newton-Raphson to make sure we converge if there is a solution.

**Reply by HugoGranstrom** [reply]
It will be slower though but hopefully not too much

**Reply by HugoGranstrom** [reply]
@TheHappyKoala Would you mind finding out one of the values for `meanAnom` and `e` that gives Nan? 😄 Then I could try it out and see if I find any algorithm that converges in another language so I don't have to write it myself? 🤣

**Reply by TheHappyKoala** [reply]
Sure, give me a couple of mins... Checking what the value of E is when entering another SOI works

**Reply by TheHappyKoala** [reply]
Mean anomaly: -6305.985121061316 Eccentricity: 1.0001823673423722

**Reply by HugoGranstrom** [reply]
Thank you! 😃 I'll go to bed now but I'll take a closer look into this tomorrow

**Reply by HugoGranstrom** [reply]
A quick-fix that may work is changing this line https://github.com/TheHappyKoala/Harmony-of-the-Spheres/blob/92aac452c832f8dab4228be0d7f664be0e7a9df6/src/physics/spacecraft/lambert.ts#L549 to:
```let E = 0```
Because we can't use meanAnom as the initial condition 'cause it's so big.

**Reply by TheHappyKoala** [reply]
Sleep tight! I'll give that a try now 😎

**Reply by TheHappyKoala** [reply]
Noops... Gonna give zzzzz a try now... Work tomorrow 🤪

**Reply by HugoGranstrom** [reply]
What a bummer! 🙃 I've found that if I try a better initial guess it does indeed converge for your values. Hopefully, that should be enough to solve this. I'll send you a snippet to try out after breakfast 😄

**Reply by HugoGranstrom** [reply]
@TheHappyKoala Here you go, I hope it works. It didn't break Lunar Free Return at least 🤣 
 
```javascript
function solveKeplerEq(meanAnom: number, e: number, tol=1e-14): number {
  // Solves Keplers third law M = E - e*sin(E) for E using Newton-Raphson
  let E = meanAnom;
  let E_last = E + 1e3*tol; // make sure we enter the while loop
  if (e < 1) {
    while (Math.abs(E - E_last) > tol) {
      E_last = E;
      E = E - (E - e * Math.sin(E) - meanAnom) / (1 - e * Math.cos(E));
    }
  } else if (e > 1) {
    let s = Math.sign(meanAnom);
    let E0 = 1e-3*s;
    while (Math.sign(meanAnom - e * Math.sinh(E0) + E0) == s) { // Find a sign change and use it as inital value
      E0 += s;
    }
    E = E0;
    E_last = E + 1e3*tol; // make sure we enter the while loop
    while (Math.abs(E - E_last) > tol) {
      E_last = E;
      E = E + (meanAnom - e * Math.sinh(E) + E) / (e * Math.cosh(E) - 1);
    }
  }
  return E
}
```

**Reply by TheHappyKoala** [reply]
Sorry for getting back to you at this late hour (first work then two hours of cooking Mie Goreng), but I'm gonna give it a spin now 😄

**Reply by HugoGranstrom** [reply]
No worries 😁 in terms of Github speed I'd say you're pretty fast 😎 many take much more time to answer hehe (I'm maybe guilty sometimes 🤣)

**Reply by TheHappyKoala** [reply]
Mars is orbiting retrograde, Mercury is orbiting more slowly than Venus and wow, what a funky Universe 🤣

**Reply by TheHappyKoala** [reply]
Funny thing is that the orbits have the correct shape

**Reply by HugoGranstrom** [reply]
What?!? 🤣🤣🤣 I thought I kept the elliptical orbits logic the same as before 🙈

**Reply by TheHappyKoala** [reply]
WAIT

**Reply by TheHappyKoala** [reply]
In all likelihood I am the clown of the 21st century

**Reply by TheHappyKoala** [reply]
Yups, I'm the clown of the 21st century and you're a champion: it worked 😄 🚀 !!!

**Reply by TheHappyKoala** [reply]
Thanks, Hugo... I've been trying to sort this out on and off since the start of the year 🤣

**Reply by TheHappyKoala** [reply]
Mega spaceflight update coming soon enough

**Reply by HugoGranstrom** [reply]
Woooohoooooo 🥳🥳🥳 Great that it worked!!! 

Looking forward to that update! 😁

**Comment by TheHappyKoala** [comment]
Will get back to you in a bit...

**Comment by TheHappyKoala** [comment]
<img width="1440" alt="Screenshot 2021-03-02 at 22 12 52" src="https://user-images.githubusercontent.com/23419647/109715971-8a4bab80-7ba4-11eb-99a8-4f70ce707fdc.png">

This is so cool!

**Comment by TheHappyKoala** [comment]
<img width="1439" alt="Screenshot 2021-03-02 at 22 14 58" src="https://user-images.githubusercontent.com/23419647/109716327-07772080-7ba5-11eb-98b4-fb5a55b5bb9c.png">
<img width="1425" alt="Screenshot 2021-03-02 at 22 16 46" src="https://user-images.githubusercontent.com/23419647/109716339-0c3bd480-7ba5-11eb-81ac-52557dcd239d.png">

**Comment by TheHappyKoala** [comment]
It's freaking spot on... We can reach the center of even the tiniest of asteroids!!!! 🚀

**Comment by TheHappyKoala** [comment]
Got some tiny improvements to make, but people are gonna have fun with this stuff

**Comment by HugoGranstrom** [comment]
Man! That's soooo cool!!! 🤩🤩🚀 

So we started in low Earth orbit and took us all the way to Ceres all automatically??!! 🤯 

How much logic have you had to implement beyond the Lambert solver to make all this work? It seems like you have put a great deal of time and effort into this 😄

**Comment by TheHappyKoala** [comment]
Indeed 🚀 ! 

In that particular case I'd deleted the Earth, since it perturbs the trajectory of the spacecraft. I'm gonna solve this issue by simply adding a trajectory correction when you leave the sphere of influence of the body the spacecraft is departing from. It's not a big correction, and I think it makes more sense than introducing hokey pokey logic where you set the mass of the departing body to be 0 as far as the spacecraft is concerned when a trajectory burn has been applied. Open to suggestions, though! 

But other than that, yes, it's a straight trip to your destination 😃 ! Even Pluto! 

There's not that much logic, but for it to be an enjoyable experience, you need quite the UI, such as the trajectory lines of both the spacecraft, planets and moons, the rendevouz label, the UI for manual rocket thrusting, time step controls and so on. There's so much stuff that one can do to make it a truly awesome and interactive experience. In terms of logic, I guess the one contribution I've made is aligning the velocity vector with the orientation of the spacecraft, which you set with the arrow buttons in the screenshot. Nothing big, but had to be done 🤣 . 

For the upcoming release I want to include hyperbolic trajectory lines, peri and apoapsis labels, a proper time step control and some more stats... The bastard will be the hyperbolic trajectory lines, but it won't compare in complexity to anything you've done 🤣 .

Had any more fun with your scope????

**Comment by TheHappyKoala** [comment]
One more thing I'll have to implement is an arrow indicating the direction vector of the spacecraft since that will make orbital insertion maneuvers much easier (basically turn the spacecraft around so that it faces in the opposite direction of the direction vector and blast the thrusters until orbit has been achieved)... Bla, bla, bla 🥱

**Comment by TheHappyKoala** [comment]
Wouldn't it be cool if the lambert solver gave you the direction vector and velocity needed to go where you want to go and then you have to do the thrusting yourself??? So it would be more of a planning tool...

**Comment by HugoGranstrom** [comment]
Ah ok, so not quite as cool then 😜 (Joking still so cool 🤩). Yes I agree, that seems to be the most efficient and easy way of doing it 😄 Btw when going to Pluto, did you do any corrections mid-flight or was it just a single thrust in the beginning? 

Yeah the UX is the single most important feature for drawing in and keeping users. No matter how good it plans the trip, if it's a disaster to control it no one will use it so that's very important work you are doing 😄 Rotations can be really confusing 😮 

Oh lord, hyperbolic trajectory lines sound like a pain in the anti-mouth 🤣 I just ported some code but I didn't have to deal with thinking about things like this so I'd argue it's a bit more complex 😉 

I haven't been out much at all during the winter. It was so cold and snowy so I didn't bother but yesterday I was out looking and I think I saw some star clusters and perhaps, just perhaps the crab nebula. 

Oh yes that arrow would be really helpful 😄 I like the idea of having an additional mode where you just get the planning tools but have to maneuver it yourself. (with or without fuel budget)

**Comment by TheHappyKoala** [comment]
No midflight corrections once you are in the Sun's sphere of influence 🚀 !  

As for the hyperbolic trajectories, I will probably end up borrowing some code from an open source repository, but before I do that I'm gonna give it a go myself. Porting code aint easy, though, so I'm still inclined to say you've done a tremendous job and you clearly understand the code since the other day you helped me fix a bug that I'd been struggling with for two months 😆 !!! 

Gonna deploy something this weekend, though, and then implement improvements incrementally, but now I'm gonna make samosas... I've become somewhat obsessed with cooking; it's so much fun, and feels kind of meditative

**Comment by HugoGranstrom** [comment]
Oh, that's impressive! 🤯 🚀

Yeah it's no shame in porting code but to write it yourself from scratch is really satisfactory 😄 Had the bug been anywhere in the Lambert solver I can guarantee you I wouldn't have been able to reason myself to find it 🤣 So we were lucky it was in the propagating code instead. 

Can't wait!!! ⭐ 🚀 🤩 Yeah cooking can be really nice when you do it because you want to and not only because you have to 😃

**Comment by TheHappyKoala** [comment]
Just pushed an update... Should be online in half an hour (the time it takes for Netlify to build it). Still early days and the UI is still crap, but you can zoom around the inner solar system and admire planets with mega high-res textures. Gonna be bucket loads of fun to introduce more features!!! 🚀 🚀 🚀 🚀 🚀 🚀 🚀 🚀

**Comment by TheHappyKoala** [comment]
<img width="1440" alt="Screenshot 2021-03-07 at 22 59 55" src="https://user-images.githubusercontent.com/23419647/110256389-f4939000-7f98-11eb-9e09-0754c4805556.png">

**Comment by TheHappyKoala** [comment]
<img width="1437" alt="Screenshot 2021-03-07 at 23 39 49" src="https://user-images.githubusercontent.com/23419647/110257483-71753880-7f9e-11eb-8b74-97f5291f7031.png">

Space spear!

**Comment by HugoGranstrom** [comment]
That was so cool!!! 🤩 🚀🚀🚀🚀

It took me a while to get it working though 😅 I waited too long to switch down the time step so I missed my target the first couple of times but it was spot on! 😲 When I finally succeeded, I managed to get in a rather circular orbit around Mars. So my first impressions was amazing 😄 But there are of course thing to improve, and mostly regarding the time steps in my opinion. First thing was that I overshot a fair number of times because I was using the largest timestep. I want to remember you proposing to lower the timestep automatically when entering the target's sphere of influence and that would be really convenient 😃 Second, I set my time of flight to the lowest and still on highest speed it took me quite a while of waiting until I arrived at my destination. So maybe even higher timesteps would be a good idea? I like the idea of having discrete buttons for specific speeds though :D Maybe they could have witty labels as well so one can get a hunch for which to choose?

Regardless of these minor issues it was a thrill to use! Great work Darrell! 🚀

**Comment by TheHappyKoala** [comment]
Cheers for your feedback! All makes an awful lot of sense and I'll implement it for the next iteration... I have an awful lot of ideas 🚀 🚀🚀🚀🚀🚀🚀!!! 

Got this from an American teacher yesterday "These are fabulous! Intend to share with students - thanks for making. The center of mass and different settings really give more insight into gravity. How accurate are the numbers such as speed, etc?"... I don't know about you, but that's why I'm doing this (and for the kick of figuring things out, of course 😉 )!!! He had some feedback, as well, so I'm gonna implement it and get back to him.

**Reply by HugoGranstrom** [reply]
Awesome! 😄 This is the best mode of programming, when you just can realize your ideas without that much friction!🚀🚀 

That is such a thrill to get those messages!🤩 It makes it totally worth it!!!😃🚀🚀
Out of curiosity, how accurate did you say the numbers were? 🤣

**Reply by TheHappyKoala** [reply]
I haven't replied yet, but I'm just gonna point him to our readme file: https://github.com/TheHappyKoala/Harmony-of-the-Spheres

**Comment by TheHappyKoala** [comment]
We get quite a lot of messages, actually.... Mostly people requesting new scenarios, but some also have really cool ideas for new features

**Comment by TheHappyKoala** [comment]
How's your Thursday treating you?

**Reply by TheHappyKoala** [reply]
Wednesday... Damn every day is the same so I loose track of the days phenomena 🤣 🙄

**Comment by HugoGranstrom** [comment]
That's nice! 😄 What kinds of scenarios are they requesting? Exoplanets or solar system?

My Thursday is treating me well, been out for a 2.5 hour walk before lunch. You must be psychic to know that I wouldn't respond yesterday 🤣🤣🤣  How is your Thursday/Friday? 😉

**Comment by HugoGranstrom** [comment]
@TheHappyKoala How is your summer? 😁

**Comment by FalloutScript** [comment]
@TheHappyKoala You should add the gas giants and the moon in the starship scenario

**Reply by TheHappyKoala** [reply]
Hello, @FalloutScript... Indeed I should; any chance you want to create an issue?

**Comment by jvanv62** [comment]
Hi there!

I was hoping to join in here, to restart my programming and to answer some questions that my reading of space opera has generated.

Last time I did much development, I was working with gcc, make, lex, and yacc, so npm, gatsby, and react are a lot to learn all at once.

And I still haven't quite figured out push and pull in git - seems like they should switch names as far as I'm concerned.  

Anyways, hope to have more substance to talk about as things progress.

-John

**Reply by TheHappyKoala** [reply]
Hi John!

Thanks for writing. We would love it if more people contributed to our simulator, so you are more than welcome to join forces 😄 . Fortunately, you don't have to know all of those technologies to contribute to Harmony of the Spheres. @HugoGranstrom, for instance,  probably hasn't touched Gatsby, React and the related libraries at all, but he's written some cracking physics code. What kind of contribution would you be interested in making? If you let me know, I can create an issue and assign it to you, and give some pointers regarding what to read up on.

Once again, many thanks for writing!

**Comment by jvanv62** [comment]
At this point I would like to focus on the user interface and the starship simulation. I read a lot of sci-fi, and I keep running into, “I don’t think that is how it would work. My physics and math are adequate, but definitely not my strongest suit. I also spent some time trying to get a phd in software engineering, so I would be happy to comment or refactor anywhere you think that would help. 

-John Van Voorhis

**Comment by HugoGranstrom** [comment]
@jvanv62 

As @TheHappyKoala said, you don't have to learn everything to help out. I for one have never touched any React- or Gatsby-code. But from what you say about working with UI I would guess React is the thing you should focus on learning at first. 

> And I still haven't quite figured out push and pull in git - seems like they should switch names as far as I'm concerned.

When you push, you push your local changes to the git server. When you pull, you pull the latest changes from the git server. 
I find the term "Pull Request" a bit confusing though. As a contributor, I am pushing my changes to a PR, not pulling them. But I am requesting the maintainer to **pull** my changes into their repository. That's where the term comes from. Hope it cleared up something at least.

**Comment by TheHappyKoala** [comment]
Hey @jvanv62 !

If you're not sure about how to use git, I would suggest using [GitHub Desktop](https://desktop.github.com/), which makes creating a new branch, pulling the latest changes from master and pushing commits super easy. 

And super sweet that you want to work on the Starship scenarios, as they haven't received the love that they so deserve. Is there something you miss in there, or something you would change? If so, I think that would be an awesome first contribution, and please don't hesitate to ask questions. 

I'm gonna clean up and improve the local development setup, since it's somewhat of a mess at the moment and could be optimised a great deal, but tell me, have you been able to install all the dependencies and start the project on your computer?

**Comment by jvanv62** [comment]
Darrell,

Actually having some trouble setting things up.  I have home-brew and node set up on my computer, but  am having trouble getting libraries/modules set up properly for HotS.  I successfully created the hello world react app, so I have everything I need somewhere.  

I have found some good online help stuff, but I think there may be some issues with out of date libraries.

Additionally, I can make git work, just not feeling comfortable with the workflow yet.

I’m running on a recent, not new, iMac with Monterery for what it’s worth.

-John

**Reply by TheHappyKoala** [reply]
Hey John, I'm gonna take care of this tomorrow and add the steps for setting things up to the readme file. There are indeed a couple of packages that need to be updated, but I don't think that's what is causing your issue.

If you run node --version from the root of the project, what does it say?

**Reply by jvanv62** [reply]
I've got 18.7.0 installed

I solved the react problem, and am working on the gatsby install.  I think I am just a lot rustier than I thought.  Using a non-admin account for development, and having to remember where I can and can't write is a pain, but helps with security.  I am pretty sure that most of the problem is my environment.

Will let you know when I succeed.

**Reply by jvanv62** [reply]
I have successs,  and it was related to the module versions. 'npm update fix' solved my immediate problems.

**Reply by TheHappyKoala** [reply]
Happy September, @jvanv62! Just want to check in and see if you have any questions or suggestions. If you want me to create and assign you an issue related to the starship scenarios, let me know and I'll get to it :).

**Comment by jvanv62** [comment]
That would be great. I’ve been busy with other stuff, but specific goals might focus my mind. 

-John Van Voorhis

**Reply by TheHappyKoala** [reply]
Happy Saturday! Yea, I know what you're talking about; work's been eating up a great deal of my time over the past few months! Here's an issue that I created: https://github.com/TheHappyKoala/Harmony-of-the-Spheres/issues/168. Think you might enjoy working on it, and if you wanna collaborate on it, let me know! I'm gonna do some starship scenario related work myself.

**Comment by Eclipse558** [comment]
can you add some new mass templates, like Sagittarius A* and other black holes? i'm a bit tired of beetlejuice being the biggest template, and i'm dumb, so i can't make it myself. thanks.

**Comment by Eclipse558** [comment]
also, did you make it the website format, or was that someone else? link: https://gravitysimulator.org/

**Comment by Eclipse558** [comment]
sorry about any future message spamming, ADHD brain sucks.

**Comment by TheHappyKoala** [comment]
Hi @Eclipse558!

No need to apologise! 

I'm rewriting the simulator, and while doing so, adding some improvements, and more mass templates, as well as more convenient controls for creating new masses from scratch, are part of these improvements. 

As for the address of the website - that's a legacy of how I set up the DNS configuration for the website server, so I take full responsibility for that 🤣 ! 

Hope all is good and sorry for the delayed response.
